### PR TITLE
Fix #23: Diffrent RPC & XML fields for EX & QFX series

### LINKED ISSUE #23: get_mac_address_table doesn't work with JunOS 12.3R6.6
get_mac_address_table doesn't seem to work with JunOS 12.3R6.6 while get_facts for example is working well:

```
>>> from napalm import get_network_driver
>>> driver = get_network_driver('junos')
>>> device = driver('10.0.0.1', '<username_removed>', '<password_removed>')
>>> device.open()
>>> print device.get_facts()
{'os_version': u'12.3R6.6', 'uptime': 56577120, 'interface_list': ['ge-0/0/0', 'ge-0/0/20', 'ge-0/0/21', 'ge-0/0/22', 'ge-0/0/23', 'xe-0/1/0', 'xe-0/1/1', 'ge-1/0/0', 'ge-1/0/1', 'ge-1/0/2', 'ge-1/0/3', 'ge-1/0/4', 'ge-1/0/5', 'ge-1/0/6', 'ge-1/0/7', 'ge-1/0/8', 'ge-1/0/9', 'ge-1/0/10', 'ge-1/0/11', 'ge-1/0/12', 'ge-1/0/13', 'ge-1/0/14', 'ge-1/0/15', 'ge-1/0/16', 'ge-1/0/17', 'ge-1/0/18', 'ge-1/0/19', 'ge-1/0/20', 'ge-1/0/21', 'ge-1/0/22', 'ge-1/0/23', 'xe-1/1/0', 'xe-1/1/1', '.local.', 'vcp-0', 'vcp-1', 'ae0', 'ae1', 'ae2', 'bme0', 'dsc', 'gre', 'ipip', 'lo0', 'lsi', 'me0', 'mtun', 'pimd', 'pime', 'tap', 'vlan', 'vme'], 'vendor': u'Juniper', 'serial_number': u'CL0201230451', 'model': u'Virtual Chassis', 'hostname': u'labcorertr01', 'fqdn': u'labcorertr01.company.com'}
>>> print device.get_mac_address_table()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/napalm_junos/junos.py", line 860, in get_mac_address_table
    mac_table.get()
  File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/factory/optable.py", line 63, in get
    self.xml = getattr(self.RPC, self.GET_RPC)(**rpc_args)
  File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/rpcmeta.py", line 143, in _exec_rpc
    return self._junos.execute(rpc, **dec_args)
  File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/decorators.py", line 58, in wrapper
    result = function(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/decorators.py", line 26, in wrapper
    return function(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/device.py", line 570, in execute
    raise e(cmd=rpc_cmd_e, rsp=rsp, errs=err)
jnpr.junos.exception.RpcError: RpcError(severity: error, bad_element: get-bridge-mac-table, message: error: syntax error
error: syntax error)
```

### PR DESCRIPTION
**Pull request by mirceaulinic**
@kaage: Try installing the new version using `pip install -e git+github.com:mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos` and lemme know if any improvement.

### CODE REVIEW

**Comment by kaage** [comment]
Cannot get new version installed:

```
user@user-virtual-machine:~/ansible_playbooks$ sudo pip install -e git+github.com:mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
Obtaining napalm-junos from git+github.com:mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
  Cloning github.com:mirceaulinic/napalm-junos.git (to CF-MAC-FIX-23) to ./src/napalm-junos
w^Hssh: connect to host github.com port 22: Connection timed out
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
  Complete output from command /usr/bin/git clone -q github.com:mirceaulinic/napalm-junos.git /home/user/ansible_playbooks/src/napalm-junos:

----------------------------------------
Cleaning up...
Command /usr/bin/git clone -q github.com:mirceaulinic/napalm-junos.git /home/user/ansible_playbooks/src/napalm-junos failed with error code 128 in None
Storing debug log for failure in /home/user/.pip/pip.log
```

**Comment by dbarrosop** [comment]
@kaage try:

`pip install -e git+https://github.com/mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos`

**Comment by kaage** [comment]
I got some problems:

```
user@user-virtual-machine:~$ sudo pip install -e git+https://github.com/mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
[sudo] password for user: 
Obtaining napalm-junos from git+https://github.com/mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
  Cloning https://github.com/mirceaulinic/napalm-junos.git (to CF-MAC-FIX-23) to ./src/napalm-junos
  Running setup.py (path:/home/user/src/napalm-junos/setup.py) egg_info for package napalm-junos

    warning: no files found matching 'napalm_junos/utils/textfsm_templates/*.tpl'
  Installing extra requirements: 'egg'
Requirement already satisfied (use --upgrade to upgrade): napalm-base in /usr/local/lib/python2.7/dist-packages (from napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): junos-eznc in /usr/local/lib/python2.7/dist-packages (from napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): jinja2 in /usr/lib/python2.7/dist-packages (from napalm-base->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): gtextfsm in /usr/local/lib/python2.7/dist-packages (from napalm-base->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): lxml>=3.2.4 in /usr/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): ncclient>=0.4.6 in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Downloading/unpacking paramiko>=1.15.2 (from junos-eznc->napalm-junos)
  Downloading paramiko-2.0.0-py2.py3-none-any.whl (170kB): 170kB downloaded
Requirement already satisfied (use --upgrade to upgrade): scp>=0.7.0 in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): PyYAML>=3.10 in /usr/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): netaddr in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): markupsafe in /usr/lib/python2.7/dist-packages (from jinja2->napalm-base->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): setuptools>0.6 in /usr/lib/python2.7/dist-packages (from ncclient>=0.4.6->junos-eznc->napalm-junos)
Downloading/unpacking pyasn1>=0.1.7 (from paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading pyasn1-0.1.9-py2.py3-none-any.whl
Downloading/unpacking cryptography>=1.1 (from paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading cryptography-1.3.2.tar.gz (383kB): 383kB downloaded
  Running setup.py (path:/tmp/pip_build_root/cryptography/setup.py) egg_info for package cryptography

    no previously-included directories found matching 'docs/_build'
    warning: no previously-included files matching '*' found under directory 'vectors'
Downloading/unpacking idna>=2.0 (from cryptography>=1.1->paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading idna-2.1-py2.py3-none-any.whl (54kB): 54kB downloaded
Requirement already satisfied (use --upgrade to upgrade): six>=1.4.1 in /usr/lib/python2.7/dist-packages (from cryptography>=1.1->paramiko>=1.15.2->junos-eznc->napalm-junos)
Downloading/unpacking enum34 (from cryptography>=1.1->paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading enum34-1.1.6-py2-none-any.whl
Downloading/unpacking ipaddress (from cryptography>=1.1->paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading ipaddress-1.0.16-py27-none-any.whl
Downloading/unpacking cffi>=1.4.1 (from cryptography>=1.1->paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading cffi-1.6.0.tar.gz (397kB): 397kB downloaded
  Running setup.py (path:/tmp/pip_build_root/cffi/setup.py) egg_info for package cffi
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found

Downloading/unpacking pycparser (from cffi>=1.4.1->cryptography>=1.1->paramiko>=1.15.2->junos-eznc->napalm-junos)
  Downloading pycparser-2.14.tar.gz (223kB): 223kB downloaded
  Running setup.py (path:/tmp/pip_build_root/pycparser/setup.py) egg_info for package pycparser

    warning: no previously-included files matching 'yacctab.*' found under directory 'tests'
    warning: no previously-included files matching 'lextab.*' found under directory 'tests'
    warning: no previously-included files matching 'yacctab.*' found under directory 'examples'
    warning: no previously-included files matching 'lextab.*' found under directory 'examples'
Installing collected packages: napalm-junos, paramiko, pyasn1, cryptography, idna, enum34, ipaddress, cffi, pycparser
  Running setup.py develop for napalm-junos

    warning: no files found matching 'napalm_junos/utils/textfsm_templates/*.tpl'
    Creating /usr/local/lib/python2.7/dist-packages/napalm-junos.egg-link (link to .)
    Adding napalm-junos 0.2.1 to easy-install.pth file

    Installed /home/user/src/napalm-junos
  Found existing installation: paramiko 1.10.1
    Not uninstalling paramiko at /usr/lib/python2.7/dist-packages, owned by OS
  Running setup.py install for cryptography
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found
    c/_cffi_backend.c:2:20: fatal error: Python.h: No such file or directory
     #include <Python.h>
                        ^
    compilation terminated.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip_build_root/cryptography/setup.py", line 335, in <module>
        **keywords_with_side_effects(sys.argv)
      File "/usr/lib/python2.7/distutils/core.py", line 111, in setup
        _setup_distribution = dist = klass(attrs)
      File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 239, in __init__
        self.fetch_build_eggs(attrs.pop('setup_requires'))
      File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 264, in fetch_build_eggs
        replace_conflicting=True
      File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 620, in resolve
        dist = best[req.key] = env.best_match(req, ws, installer)
      File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 858, in best_match
        return self.obtain(req, installer) # try and download/install
      File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 870, in obtain
        return installer(requirement)
      File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 314, in fetch_build_egg
        return cmd.easy_install(req)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 616, in easy_install
        return self.install_item(spec, dist.location, tmpdir, deps)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 646, in install_item
        dists = self.install_eggs(spec, download, tmpdir)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 834, in install_eggs
        return self.build_and_install(setup_script, setup_base)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1040, in build_and_install
        self.run_setup(setup_script, setup_base, args)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1028, in run_setup
        raise DistutilsError("Setup script exited with %s" % (v.args[0],))
    distutils.errors.DistutilsError: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-nJLEiH-record/install-record.txt --single-version-externally-managed --compile:
    Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

Package libffi was not found in the pkg-config search path.

Perhaps you should add the directory containing `libffi.pc'

to the PKG_CONFIG_PATH environment variable

No package 'libffi' found

c/_cffi_backend.c:2:20: fatal error: Python.h: No such file or directory

 #include <Python.h>

                    ^

compilation terminated.

Traceback (most recent call last):

  File "<string>", line 1, in <module>

  File "/tmp/pip_build_root/cryptography/setup.py", line 335, in <module>

    **keywords_with_side_effects(sys.argv)

  File "/usr/lib/python2.7/distutils/core.py", line 111, in setup

    _setup_distribution = dist = klass(attrs)

  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 239, in __init__

    self.fetch_build_eggs(attrs.pop('setup_requires'))

  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 264, in fetch_build_eggs

    replace_conflicting=True

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 620, in resolve

    dist = best[req.key] = env.best_match(req, ws, installer)

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 858, in best_match

    return self.obtain(req, installer) # try and download/install

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 870, in obtain

    return installer(requirement)

  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 314, in fetch_build_egg

    return cmd.easy_install(req)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 616, in easy_install

    return self.install_item(spec, dist.location, tmpdir, deps)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 646, in install_item

    dists = self.install_eggs(spec, download, tmpdir)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 834, in install_eggs

    return self.build_and_install(setup_script, setup_base)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1040, in build_and_install

    self.run_setup(setup_script, setup_base, args)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1028, in run_setup

    raise DistutilsError("Setup script exited with %s" % (v.args[0],))

distutils.errors.DistutilsError: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-nJLEiH-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip_build_root/cryptography
Storing debug log for failure in /home/user/.pip/pip.log


user@user-virtual-machine:~/ansible_playbooks$ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from napalm import get_network_driver
>>> driver = get_network_driver('junos')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/napalm_base/__init__.py", line 73, in get_network_driver
    install_name=module_install_name
napalm_base.exceptions.ModuleImportError: Cannot import "napalm_junos". Is the library installed?
>>> 

```

**Comment by dbarrosop** [comment]
That's some third party library failing to install. I think you need to install with apt-get/yum `libpython2.7-dev` or whatever it's called in your distro.

**Comment by kaage** [comment]
I was able to install libpython2.7-dev and did following after that:

```
user@user-virtual-machine:~/ansible_playbooks$ sudo pip install -e git+https://github.com/mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
Obtaining napalm-junos from git+https://github.com/mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
  Updating ./src/napalm-junos clone (to CF-MAC-FIX-23)
  Running setup.py (path:/home/user/ansible_playbooks/src/napalm-junos/setup.py) egg_info for package napalm-junos

    warning: no files found matching 'napalm_junos/utils/textfsm_templates/*.tpl'
  Installing extra requirements: 'egg'
Requirement already satisfied (use --upgrade to upgrade): napalm-base in /usr/local/lib/python2.7/dist-packages (from napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): junos-eznc in /usr/local/lib/python2.7/dist-packages (from napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): jinja2 in /usr/lib/python2.7/dist-packages (from napalm-base->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): gtextfsm in /usr/local/lib/python2.7/dist-packages (from napalm-base->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): lxml>=3.2.4 in /usr/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): ncclient>=0.4.6 in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): paramiko>=1.15.2 in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): scp>=0.7.0 in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): PyYAML>=3.10 in /usr/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): netaddr in /usr/local/lib/python2.7/dist-packages (from junos-eznc->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): markupsafe in /usr/lib/python2.7/dist-packages (from jinja2->napalm-base->napalm-junos)
Requirement already satisfied (use --upgrade to upgrade): setuptools>0.6 in /usr/lib/python2.7/dist-packages (from ncclient>=0.4.6->junos-eznc->napalm-junos)
Installing collected packages: napalm-junos
  Running setup.py develop for napalm-junos

    warning: no files found matching 'napalm_junos/utils/textfsm_templates/*.tpl'
    Creating /usr/local/lib/python2.7/dist-packages/napalm-junos.egg-link (link to .)
    napalm-junos 0.2.1 is already the active version in easy-install.pth

    Installed /home/user/ansible_playbooks/src/napalm-junos
Successfully installed napalm-junos
Cleaning up...
user@user-virtual-machine:~/ansible_playbooks$ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from napalm import get_network_driver
>>> driver = get_network_driver('junos')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/napalm_base/__init__.py", line 73, in get_network_driver
    install_name=module_install_name
napalm_base.exceptions.ModuleImportError: Cannot import "napalm_junos". Is the library installed?

```

**Comment by dbarrosop** [comment]
Try this:

```
sudo pip uninstall -y napalm_junos
sudo pip install -e git+https://github.com/mirceaulinic/napalm-junos.git@CF-MAC-FIX-23#egg=napalm-junos
```

**Comment by kaage** [comment]
```
user@user-virtual-machine:~/ansible_playbooks$ sudo pip uninstall -y napalm_junos
Can't uninstall 'napalm-junos'. No files were found to uninstall.
```

**Comment by mirceaulinic** [comment]
It seems that not only the RPC is different but also the XML reply:

MX series:

``` xml
<rpc-reply xmlns:junos="http://xml.juniper.net/junos/13.3R6/junos">
    <l2ald-rtb-macdb>
        <l2ald-mac-entry junos:style="extensive">
            <l2-mac-address>00:11:22:33:44:55</l2-mac-address>
            <l2-mac-routing-instance>default-switch</l2-mac-routing-instance>
            <l2-mac-bridging-domain>vlan-1</l2-mac-bridging-domain>
            <l2-bridge-vlan>1</l2-bridge-vlan>
            <l2-mac-logical-interface>xe-0/0/0</l2-mac-logical-interface>
            <l2-mac-base-logical-interface>xe-0/0/0.1</l2-mac-base-logical-interface>
            <l2-mac-ifl-generation>1</l2-mac-ifl-generation>
            <l2-mac-entry-flags>in_hash,in_ifd,in_ifl,in_vlan,in_rtt,kernel,in_ifbd</l2-mac-entry-flags>
            <l2-mac-epoch>3</l2-mac-epoch>
            <l2-mac-sequence-number>2</l2-mac-sequence-number>
            <l2-mac-learn-mask>0x0000001</l2-mac-learn-mask>
        </l2ald-mac-entry>
    </l2ald-rtb-macdb>
</rpc-reply>
```

EX series:

``` xml
<rpc-reply xmlns:junos="http://xml.juniper.net/junos/12.3R6/junos">
    <ethernet-switching-table-information junos:style="extensive">
        <ethernet-switching-table junos:style="extensive">
            <mac-table-entry junos:style="extensive">
                <mac-vlan>default</mac-vlan>
                <mac-vlan-tag>0</mac-vlan-tag>
                <mac-address>aa:bb:cc:dd:ee:ff</mac-address>
                <mac-interface>Router</mac-interface>
                <mac-interfaces-list>
                    <mac-interfaces>Router</mac-interfaces>
                </mac-interfaces-list>
                <mac-type>Static</mac-type>
                <mac-action>Forward</mac-action>
                <mac-nexthop>0</mac-nexthop>
            </mac-table-entry>
    </ethernet-switching-table-information>
</rpc-reply>
```

**Comment by mirceaulinic** [comment]
This will fix #23: https://github.com/napalm-automation/napalm-junos/pull/25/commits/8a6deeeb215253439fc411de3c1f8447fa560c8b

**Comment by mirceaulinic** [comment]
@dbarrosop this can be merged and released.

File: napalm_junos/junos.py [hunk 1]
```diff
@@ -854,20 +854,38 @@ def get_interfaces_ip(self):
 
     def get_mac_address_table(self):
 
-        mac_address_table = list()
+        mac_address_table = []
 
         mac_table = junos_views.junos_mac_address_table(self.device)
+
+        if self.device.facts.get('personality', '') in ['SWITCH']:  # for EX & QFX devices
```
Comment (dbarrosop) [inline napalm_junos/junos.py:859]: Wouldn't be cleaner to have a different juniper view?

Comment (mirceaulinic) [inline napalm_junos/junos.py:859]: Hum.. maybe, but would be loaded even when the device is not EX/QFX. But would be cleaner indeed.

Comment (dbarrosop) [inline napalm_junos/junos.py:859]: Just name it differently, for example, `junos_mac_address_table_ex_qfx` or something meaningful and then:

```
if self.device.facts.get('personality', '') in ['SWITCH']:  # for EX & QFX devices
     mac_table = junos_views.junos_mac_address_table_ex_qfx(self.device)
else:
     mac_table = junos_views.junos_mac_address_table(self.device)
```

Wouldn't that work?

Comment (mirceaulinic) [inline napalm_junos/junos.py:859]: Absolutelty!

**Comment by mirceaulinic** [comment]
@dbarrosop https://travis-ci.org/napalm-automation/napalm-junos/builds/138783086#L2450 oops that's not good :(

**Comment by dbarrosop** [comment]
Yeah, I noticed that with `napalm-eos`. `pyangbind` has the same problem as napalm when installed that way. The fix is super simple, just update the travis file like this:

https://github.com/napalm-automation/napalm-eos/blob/develop/.travis.yml#L4

Can you update it, please?

**Comment by mirceaulinic** [comment]
If the changes I applied would not be correct, the tests we currently have would not catch the problem. This is because the the fake JunOS driver is a vSRX version 12.something if I recall correctly. We discussed very briefly this situation some time ago, we need test not only agains the OS but also against the firmware version (and hope that's all). @dbarrosop do you know if Matt has something like that in mind as part of his proposal? If not, we got to do it.

**Comment by dbarrosop** [comment]
I don't think so. However, I do have plans for that. I just need the time to completely refactor the testing framework. My idea is as follows (short version):

Background: For the tests we have fakedevices and mocked data. The mocked data is just a single file that fakes the output of the data of a single command.

New proposal: Have a folder per test where we can have several subfolders. Each subfolder would represent a test case. So on each test case we could have a combination of different outputs to test different cases (empty output, one single item in the output, several items, etc) and different vendors (junos12, junos13) or whatever else we have to test. This shouldn't be difficult. Just a lot to refactor. What do you think? I can try to build a POC.

**Comment by mirceaulinic** [comment]
Oh, yes, I remember! Great idea, I agree!

**Comment by mirceaulinic** [comment]
@dbarrosop I think this one is ready too. We can come back later and add fancy tests under the new framework.

**Pull request merged**